### PR TITLE
Extract GEM_NAME_RESERVED_LIST into database and include in admin panel.

### PR DESCRIPTION
**Pull request by simi**
- follow-up of https://github.com/rubygems/rubygems.org/pull/3875#issuecomment-1614308765
- potentially closes https://github.com/rubygems/rubygems/pull/6769
- I'll remove reserve action in following PR if welcomed

### CODE REVIEW

File: app/models/gem_name_reservation.rb [hunk 1]
```diff
@@ -0,0 +1,5 @@
+class GemNameReservation < ApplicationRecord
```
Comment (segiddins) [inline app/models/gem_name_reservation.rb:1]: We should validate uniqueness of the reservations?

File: app/policies/gem_name_reservation_policy.rb [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+
+  def avo_index?
+    true
+  end
+
+  def avo_show?
+    true
+  end
+
+  def avo_create?
```
Comment (segiddins) [inline app/policies/gem_name_reservation_policy.rb:16]: Create and destroy should be limited to rubygems.org admins

File: db/migrate/20230702010946_create_gem_name_reservations.rb [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+    active-support
+    active_support
+    sprockets_rails
+    rail-ties
+    rail_ties
+  ]
+
+  def change
+    create_table :gem_name_reservations do |t|
+      t.string :name
```
Comment (segiddins) [inline db/migrate/20230702010946_create_gem_name_reservations.rb:59]: Unique index

File: app/policies/gem_name_reservation_policy.rb (hunk 2 above)
Comment (simi) [reply app/policies/gem_name_reservation_policy.rb:16]: Thanks, I have updated this. I was looking into other policies to inspire. Is there currently any difference in this behaviour?

File: db/migrate/20230702010946_create_gem_name_reservations.rb (hunk 3 above)
Comment (simi) [reply db/migrate/20230702010946_create_gem_name_reservations.rb]: done, I have also added `null: false`

File: app/models/gem_name_reservation.rb (hunk 1 above)
Comment (simi) [reply app/models/gem_name_reservation.rb:1]: added

**Comment by hsbt** [comment]
👍  This feature is super helpful for me.

**Review by mensfeld — APPROVED** [review]

File: app/models/gem_name_reservation.rb [hunk 4]
```diff
@@ -0,0 +1,7 @@
...
+  validates :name, uniqueness: { case_sensitive: false }, presence: true
```
Comment (segiddins) [inline app/models/gem_name_reservation.rb:2]: nit: also validate that `name == name.downcase` ?

**Review by segiddins — APPROVED** [review]
Nice!

File: app/models/gem_name_reservation.rb (hunk 4 above)
Comment (simi) [reply app/models/gem_name_reservation.rb:2]: updated

**Comment by segiddins** [comment]
@simi feel free to deploy whenever you're ready!

**Pull request merged**